libs/SalesforceSDK/src/com/salesforce/androidsdk/security/KeyStoreWrapper.java

```diff
@@ -0,0 +1,188 @@
+ * and/or other materials provided with the distribution.
+ * - Neither the name of salesforce.com, inc. nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission of salesforce.com, inc.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.security;
+
+import android.os.Build;
+import android.security.KeyPairGeneratorSpec;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+import android.util.Base64;
+
+import com.salesforce.androidsdk.app.SalesforceSDKManager;
+import com.salesforce.androidsdk.util.SalesforceSDKLogger;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.CertificateException;
+import java.util.Calendar;
+
+import javax.security.auth.x500.X500Principal;
+
+/**
+ * This class provides utilities to interact with the Android KeyStore.
+ * For more information on the KeyStore, see {@link KeyStore}.
+ *
+ * @author bhariharan
+ */
+public class KeyStoreWrapper {
+
+    private static final String ANDROID_KEYSTORE = "AndroidKeyStore";
+    private static final String RSA = "RSA";
+    private static final String TAG = "KeyStoreWrapper";
+
+    private static KeyStoreWrapper INSTANCE;
+
+    private KeyStore keyStore;
+
+    /**
+     * Returns an instance of this class, after initializing the KeyStore if required.
+     *
+     * @return Instance of this class.
+     */
+    public synchronized static KeyStoreWrapper getInstance() {
+        if (INSTANCE == null) {
+            try {
+                INSTANCE = new KeyStoreWrapper();
+                INSTANCE.keyStore = INSTANCE.loadKeyStore();
+            } catch (Exception e) {
+                SalesforceSDKLogger.e(TAG, "Could not load KeyStore", e);
+            }
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Generates an RSA keypair and returns the public key.
```

Comment:
Could we use Elliptic Curve encryption instead of RSA: it's what we use on iOS / key generation is faster / stronger for the same number of bits  (a 256 bit ECC key is equivalent to RSA 3072 bits).

Reply:
This is existing functionality for encrypted push notifications. I will be adding methods for ECC in the next PR as I mentioned in the description.